[4.0] [New feature] Add tls transfer encryption for database connections

richard67 opened the pull request:
Pull Request for Issue # .

### Summary of Changes

Add option to global config to enable and configure data in transit encryption with TLS in MySQLi/PDO MySQL/PDO PostgreSQL drivers.

Add info about this encryption to sysinfo and privacy status module.

Update database framework package to the latest version which supports this feature.

Thanks to @andrepereiradasilva for the implementation the support for this feature in the database framework (PR's [https://github.com/joomla-framework/database/pull/177](https://github.com/joomla-framework/database/pull/177) and [https://github.com/joomla-framework/database/pull/183](https://github.com/joomla-framework/database/pull/183)) and for almost all of the code for this PR.

### Testing Instructions

**Pre-condition:**  Set up your database server to support encrypted connections.

Links to some helpful documentation:

- MySQL 5.7 [https://dev.mysql.com/doc/mysql-security-excerpt/5.7/en/encrypted-connections.html](https://dev.mysql.com/doc/mysql-security-excerpt/5.7/en/encrypted-connections.html)
- MySQL 8.0 [https://dev.mysql.com/doc/refman/8.0/en/encrypted-connections.html](https://dev.mysql.com/doc/refman/8.0/en/encrypted-connections.html)
- MariaDB [https://mariadb.com/kb/en/library/data-in-transit-encryption/](https://mariadb.com/kb/en/library/data-in-transit-encryption/) and [https://mariadb.com/kb/en/library/securing-connections-for-client-and-server/](https://mariadb.com/kb/en/library/securing-connections-for-client-and-server/)
- PostgreSQL 11 [https://www.postgresql.org/docs/11/ssl-tcp.html](https://www.postgresql.org/docs/11/ssl-tcp.html)
- Good explanation about certificate creation with openssl on Linux for PosgreSQL, but also useful for MySQL if on Linux: [https://www.postgresql.org/docs/11/ssl-tcp.html#SSL-CERTIFICATE-CREATION](https://www.postgresql.org/docs/11/ssl-tcp.html#SSL-CERTIFICATE-CREATION)

**Method 1 - this requires composer to be installed**

1. Apply this Pull Request (PR) on an existing installation of a clean 4.0-dev, using the database server prepared in step 1.

2. Do a `composer update joomla/database` to update the database framework package to the latest version.

3. Login to backend and change database server in global config server tab from localhost to the full hostname (fqdn) of that server and make sure it can be accessed in that way. E.g. it might be necessary to add it to your hosts file or dns for proper name to ip resolving. The reason for this is that server certificates will normally not be made for localhost.

4. Check the System Information view.

Result: See section "Expected result" below.

5. Go to Global Cofiguration, tab "Server" and check the options for database connection encryption.

Result: See section "Expected result" below.

6. Add the Privacy Status module to one of the dashboards, e.g. the "Users" dashboard, then check display of database connection encryption status in that module.

Result: See section "Expected result" below.

7. Play around with the options. Note that this SSL stiff is a bit tricky, so first read all available documentation linked above and research for your db server and server operating system before blaming this PR doing something wrong. Check the system info after changes.

Result: See section "Expected result" below.

**Method 2 - use a patched nightly built**

1. Download the nightly build package of today, October 12, 2019, patched with the changes of this PR plus the updated database package from framework here: [https://test5.richard-fath.de/Joomla_4.0.0-alpha12-dev-Development-Full_Package_2019-10-14_pr-26375.zip](https://test5.richard-fath.de/Joomla_4.0.0-alpha12-dev-Development-Full_Package_2019-10-14_pr-26375.zip).

2. Install Joomla with the package downloaded in step 1. When specifying the database server don't use localhost, use the full hostname (fqdn) of that server and make sure it can be accessed in that way. E.g. it might be necessary to add it to your hosts file or dns for proper name to ip resolving. The reason for this is that server certificates will normally not be made for localhost.

3. Check the System Information view.

Result: See section "Expected result" below.

4. Go to Global Cofiguration, tab "Server" and check the options for database connection encryption.

Result: See section "Expected result" below.

5. Add the Privacy Status module to one of the dashboards, e.g. the "Users" dashboard, then check display of database connection encryption status in that module.

Result: See section "Expected result" below.

6. Play around with the options. Note that this SSL stiff is a bit tricky, so first read all available documentation linked above and research for your db server and server operating system before blaming this PR doing something wrong. Check the system info after changes.

Result: See section "Expected result" below.

### Expected result

**System Information**

1. No encryption active:

![snap-6](https://user-images.githubusercontent.com/7413183/65379584-ac13e400-dcca-11e9-99ae-f203edd66fcb.png)

2. Encryption active:

![snap-10](https://user-images.githubusercontent.com/7413183/65379857-d5833e80-dccf-11e9-8dd7-c8940d325435.png)

**Privacy Status module**

1. No encryption active:

![db-tls-mod-priv-sts-disabled](https://user-images.githubusercontent.com/7413183/66258883-a3cba680-e7aa-11e9-8ea6-0178312831bc.png)

2. Encryption active:

![db-tls-mod-priv-sts-enabled](https://user-images.githubusercontent.com/7413183/66258888-ad550e80-e7aa-11e9-951d-1b0a334b8dcf.png)

**Global Configuration, tab "Server", section "Database Settings"**

1. Default, i.e. encryption is controlled by the server:

![snap-7](https://user-images.githubusercontent.com/7413183/65386573-d3f05f80-dd3d-11e9-885d-007de3fcf22e.png)

2. One way encryption requested by the client:

![snap-8](https://user-images.githubusercontent.com/7413183/65386577-d8b51380-dd3d-11e9-8d37-4277f78db531.png)

3. Two way encryption requested by the client:

![snap-9](https://user-images.githubusercontent.com/7413183/65386682-46157400-dd3f-11e9-941b-4eed78cd40d8.png)

### Actual result

Nothing of the above.

### Documentation Changes Required

It needs at least to extend the documentation of the server section of Global Configuration, the Privacy Dashboard Module and the System Information view by the new fields.

In addition a documentation about setting up the particular kinds of encrypted server connections could be helpful.

richard67 committed:
Add tls transfer encryption for database connections

richard67 committed:
Update composer.lock

richard67 commented:
@alikon @andrepereiradasilva @twister65 Could you test this PR?

andrepereiradasilva commented:
thanks. Will test when possible. Just some early coments:
1. "No encryption" value is not right. At least in PostgreSQL you can force encryption just in server side. So maybe "Default (server prefered)" or (better) have a new field that, for that option value, shows the current encryption status.
2. "Verify" should be "Verify Server Certificate". In 2 way connection you have a server and a client Certificate so it should be clear that in this case is the server certificate
3. IMHO whether the db connection is encrypted or not, is also a data privacy question, so this information should be also in the privacy component. If you have a database connection to other server or to a cloud database as a service, to protect privacy data should to be encrypted.
But this is alse teus for HTTPS, so maybe for other PR.

richard67 commented:
>     2. "Verify" should be "Verify Server Certificate". In 2 way connection you have a server and a client Certificate so it should be clear that, in this case, is the db server certificate that is being veryfied

@andrepereiradasilva That's why I've added this information to the "..._DESCR" language strings. But you are right, the title could be changed. Will do that, especially because it seems to me that these "..._DESCR" language strings aren't used anywhere.

Maybe @brianteeman could have a look at the language strings?

richard67 committed:
Better language strings 1

infograf768 commented:
Suggestion:
Prevent "Connexion Encryption" field to be displayed by itself facing database prefix with Verify field not facing it.

It's just a matter of adding a spacer field after the `dbprefix` field.
```
                <field
			type="spacer"
		/>
```

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -34,6 +34,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_DOMAIN_DESC="Precede domain with '.' if cookie should be valid for all subdomains."
 COM_CONFIG_FIELD_COOKIE_DOMAIN_LABEL="Cookie Domain"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The full path of the Certificate Authority (CA) certificate file."
```
### 
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The full path to the Certificate Authority (CA) certificate file."
```

richard67 committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-Authored-By: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -34,6 +34,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_LABEL="Certificate File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="The list of permissible ciphers for connection encryption as set on the server."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_LABEL="Cipher Suite"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_DESC="Data in transit encryption with TLS for database connections"
```
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_DESC="Data in transit encryption (with TLS for database connections)."
```

richard67 committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-Authored-By: Brian Teeman <brian@teeman.net>

richard67 commented:
@infograf768 Would be the first spacer in that xml, and it does not look nice on small screens (mobile). So I leave that decision for later.

richard67 commented:
@infograf768 The whole Global Settings would look better if the switchers and text fields and drop down selects all would have the same height. But that's another issue .. it just becomes very visible when I apply your suggestion so the "Verify .." switcher is right beside the "Connection Encryption" drop down on a large screen.

richard67 commented:
Thanks @brianteeman for language strings review.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

Quy commented:
I don't have step 1 configured.

Then here is the error on the System Information page:
`Call to undefined method Joomla\Database\Mysqli\MysqliDriver::isConnectionEncryptionSupported() `

richard67 commented:
@Quy Did you do step 3 composer update to get the latest version of the database framework package?

Quy commented:
Yes,

richard67 commented:
@Quy Could you check composer.lock, if it contains same version as the one in this PR for the database package?

richard67 commented:
@Quy If yes and composer.lock is ok, then it needs maybe a `composer install` instead of a `composer update joomla/database`?

Quy commented:
Yes
```
            "name": "joomla/database",
            "version": "dev-2.0-dev",
            "source": {
                "type": "git",
                "url": "https://github.com/joomla-framework/database.git",
                "reference": "c5cd726fb6137dede3a0f141df4271df310a9568"
            },
            "dist": {
                "type": "zip",
                "url": "https://api.github.com/repos/joomla-framework/database/zipball/c5cd726fb6137dede3a0f141df4271df310a9568",
                "reference": "c5cd726fb6137dede3a0f141df4271df310a9568",
                "shasum": ""
            },
```

richard67 commented:
@Quy That error normally comes only when the database package is not up to date enough to have that function, like it was for me before I did the composer update. Could you try composer install? Maybe update does nothing if composer.lock is already updated?

Quy commented:
I just did `composer install` and `composer update joomla/database` and it killed MySQL. Let me play some more and get back to you.

```
11:38:27 AM  [mysql] 	Error: MySQL shutdown unexpectedly.
11:38:27 AM  [mysql] 	This may be due to a blocked port, missing dependencies, 
11:38:27 AM  [mysql] 	improper privileges, a crash, or a shutdown by another method.
11:38:27 AM  [mysql] 	Press the Logs button to view error logs and check
11:38:27 AM  [mysql] 	the Windows Event Viewer for more clues
11:38:27 AM  [mysql] 	If you need more help, copy and post this
11:38:27 AM  [mysql] 	entire log window on the forums
```

Quy commented:
Here is from the Log:
```
2019-09-25 11:41:59 0 [Note] InnoDB: Mutexes and rw_locks use Windows interlocked functions
2019-09-25 11:41:59 0 [Note] InnoDB: Uses event mutexes
2019-09-25 11:41:59 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2019-09-25 11:41:59 0 [Note] InnoDB: Number of pools: 1
2019-09-25 11:41:59 0 [Note] InnoDB: Using SSE2 crc32 instructions
2019-09-25 11:41:59 0 [Note] InnoDB: Initializing buffer pool, total size = 16M, instances = 1, chunk size = 16M
2019-09-25 11:41:59 0 [Note] InnoDB: Completed initialization of buffer pool
2019-09-25 11:42:00 0 [Note] InnoDB: 128 out of 128 rollback segments are active.
2019-09-25 11:42:00 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2019-09-25 11:42:00 0 [Note] InnoDB: Setting file 'C:\xampp\mysql\data\ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2019-09-25 11:42:00 0 [Note] InnoDB: File 'C:\xampp\mysql\data\ibtmp1' size is now 12 MB.
2019-09-25 11:42:00 0 [Note] InnoDB: Waiting for purge to start
2019-09-25 11:42:00 0 [Note] InnoDB: 10.3.16 started; log sequence number 1143440372; transaction id 1748347
2019-09-25 11:42:00 0 [Note] InnoDB: Loading buffer pool(s) from C:\xampp\mysql\data\ib_buffer_pool
2019-09-25 11:42:00 0 [Note] InnoDB: Buffer pool(s) load completed at 190925 11:42:00
2019-09-25 11:42:00 0 [Note] Plugin 'FEEDBACK' is disabled.
2019-09-25 11:42:00 0 [Note] Server socket created on IP: '::'.
```

richard67 commented:
@Quy Ok, if necessary I'm available at Glip, too. MySQL version and MySQLi client version and so on would be good to know if problem persists. The 2nd log does not look strange, but the first one does. I have no idea yet what the problem could be, except of what the 1st log sais: " ... blocked port, missing dependencies, improper privileges ...". Maybe MySQL on Windows needs some additional SSL lib for encryption? I'll see if I can find something in the docs.

Quy commented on libraries/src/Factory.php:
```diff
@@ -667,6 +667,24 @@ protected static function createDbo()
+				'verify_server_cert' => (bool) $conf->get('dbsslverifyservercert'),
+			];
+
+			foreach (['cipher', 'ca', 'capath', 'key', 'cert'] as $key => $value)
```
```suggestion
			foreach (['cipher', 'ca', 'capath', 'key', 'cert'] as $value)
```

richard67 replied:
@Quy Funny, the same error I've corrected once before, seems it came back or it was at another place. Same error is also still in other file here, I will correct in 2 minutes.

richard67 replied:
Ah, not really an error, just an unused key ... correction on the way.

richard67 replied:
Done.

Quy commented:
The first output is from XAMPP Control Panel to launch MySQL.

richard67 committed:
Remove unused key from foreach

richard67 commented:
No idea how that could come, that a composer install or update killy MySQL on Windows.

Quy commented:
After selecting `One-way encryption` and clicking `Save`, I get a white screen with the following errors:
```
Warning: session_start(): Failed to initialize storage module: user (path: C:\xampp\tmp) in \libraries\vendor\joomla\session\src\Storage\NativeStorage.php on line 478

Warning: mysqli::prepare(): invalid object or resource mysqli in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 145

Warning: mysqli::prepare(): invalid object or resource mysqli in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 145
Error: MySQL server has gone away: Could not connect to database: MySQL server has gone away
Warning: mysqli::prepare(): invalid object or resource mysqli in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 145

Warning: session_write_close(): Failed to write session data using user defined save handler. (session.save_path: C:\xampp\tmp) in \libraries\vendor\joomla\session\src\Storage\NativeStorage.php on line 114
```

richard67 commented:
@Quy Sorry I was lazy with testing instructions. Please check description at https://github.com/joomla-framework/database/pull/177 . It’s important not to run with socket connection but with TCP/IP, and to use the host for which the certs are made, and that is possibly not localhost event if dB runs on same server as the site. It is all a tricky thing, and it seems some error handling should be added ... suggestions are welcome.

richard67 commented:
@Quy Maybe the MySQL log shows more details about what happened?

richard67 commented:
@wilsonge Maybe I should flag this PR as draft or RFC or even close it? It is more a prototype than ready for use, and I can’t do much for it because have other tasks, so it is more a kind of a starting point for people to work on it than a thing which I could make ready for use soon all alone.

andrepereiradasilva committed:
Need to check database connection on global configuration save

If not checked here the connection error on save (for instance using a unix socket in "localhost") will not be captured which in put the site down until someone manualy changes the configuration.php file.

andrepereiradasilva committed:
Add new global variables to installation config file

andrepereiradasilva committed:
more usable error message

andrepereiradasilva committed:
Update en-GB.com_config.ini

richard67 committed:
Merge pull request #7 from andrepereiradasilva/patch-33

Add new global variables to installation config file

richard67 committed:
Merge pull request #6 from andrepereiradasilva/patch-31

Need to check database connection on global configuration save

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

andrepereiradasilva committed:
Get databaseConnectionEncryption value in module start

andrepereiradasilva committed:
Add a new line showing the database connection encryption status

andrepereiradasilva committed:
Add the language variables

richard67 committed:
Merge pull request #9 from andrepereiradasilva/patch-33

Add database connection encryption status to privacy status module

richard67 committed:
CS

richard67 commented:
Whoever has interest in this feature: Could you test again? Thanks to @andrepereiradasilva error there have been made some improvements.

richard67 commented:
@brianteeman Could you review language stings again? Some new ones have been added with new commits. Thanks in advance.

twister65 commented:
I can't apply this patch with the latest nightly build:

> The file marked for modification does not exist: composer.lock

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 commented:
@twister65 That's why the testing instructions say in step 2: "Apply this Pull Request (PR) on a 4.0-dev ...", and in step 3: "Do a `composer update joomla/database`". This means you need to test on a clone of the Joomla CMS' GitHub repository on the 4.0-dev branch, and your testing environment has to be set up as described here: [https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment](https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment).

twister65 commented:
Is there a reason for this ? Because the nightly build also includes the latest joomla database framework. And I use Patchtester to test PR, so convenient.

richard67 commented:
@twister65 No, the nightly build does not include the latest state of the database framework, that's why this PR changes composer.lock to fetch the newer version. Version does not mean any release version. is is always 2.0-dev, and the commit checksum is what makes the "version". You can see that when expanding the diff for composer.lock in the list of changed files of this PR here.

twister65 commented:
Yes it does, I checked it before. PR's joomla-framework/database#177 and joomla-framework/database#183 have been merged.

SharkyKZ commented:
joomla-framework/database#183 is not included in nightly.

twister65 commented:
Exact @SharkyKZ , I'm waiting tomorrow morning then...

SharkyKZ commented:
It won't be included tomorrow either, unless someone submits a separate PR to update Composer dependencies (all or just `joomla/database`). This is already done in this PR.

richard67 commented:
@twister65 and other potential testers: I've updated my testing instructions in the description above by a 2nd test method, which uses a patched nightly build of today. You can find the link for download in the testing instructions for that method. The patched nighly build contains the changes of this PR plus the updated database package from framework.

andrepereiradasilva committed:
language clarifications

andrepereiradasilva committed:
hints and allow more characters

andrepereiradasilva committed:
Need to save in install configuation.php (on install) the new default global config values

andrepereiradasilva committed:
corrections

richard67 commented:
@twister65 and other potential testers: Please wait with testing, I will add some more changes tonight or tomorrow, German time. I'll let you know when ready.

andrepereiradasilva committed:
correct minor bug and put more friendly client certificates example path that doesn't need special permissions (open_basedir, user/group permissions) for php to access it

andrepereiradasilva committed:
Update en-GB.com_config.ini

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 committed:
Merge remote-tracking branch 'upstream/4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 committed:
Merge pull request #10 from andrepereiradasilva/patch-33

Added hints, language clarifications, and allow more characters for paths

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 commented:
@twister65 and other potential testers: I've updated the zip package for testing to the state of right now. PR is ready for review and tests.

richard67 commented:
@brianteeman Could you review the changed language stings and what is much more important, review following point regarding UX consistency? I would be really happy if you could do that.

Following is the UX consistency issue: Up to now, we don't use the `hint` property in the text fields of the Global Configuration form. We use it in a few other forms in backend, but not in Global config. The `hint` is shown in a grey color inside the field and serves as a hint or example how to fill in the field. Now this PR introduces the usage of hints in Global Config, and because the grey color enough is not really sufficient to show that the content is not a valid default value but a hint, e.g. with an example in case of this PR, the hint texts are starting with 'ex: ', which stands for example. But this has been done nowhere else in fields in backend where hints are used, no prefix for such purpose.

What do you think? Shall we drop the hints from this PR? Or change 'ex: ' to 'example: '? Or change it to 'hint: '? Or maybe make another PR which adds a static text prefix 'hint: ' (or course translatable) to all texts of those hints?

Your recommendation would be very welcome.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_DOMAIN_DESC="Precede domain with '.' if cookie should be valid for all subdomains."
 COM_CONFIG_FIELD_COOKIE_DOMAIN_LABEL="Cookie Domain"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The absolute file path to the Certificate Authority (CA) base64 encoded certificates files."
```
is it one file or multiple?

richard67 replied:
Is one file. I'll correct.

richard67 replied:
Is one file, I'll correct soon.

richard67 replied:
Done.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The absolute file path to the Certificate Authority (CA) base64 encoded certificates files."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_HINT="ex: /path/to/joomla/site/administrator/tls/ca/my-client-certificate-ca.pem"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_LABEL="CA File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="The full path to Certificate Authority (CA) base64 encoded certificate files (MySQL only). Leave it empty if \"CA File\" already have all the CA certificates needed to validate the chain of trust."
```
same question - is it singular or multiple

richard67 replied:
Here it is a folder which can contain multiple cerificate files for the complete chain of trust of root CA and intermediate CA.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="The full path to Certificate Authority (CA) base64 encoded certificate files (MySQL only). Leave it empty if \"CA File\" already have all the CA certificates needed to validate the chain of trust."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_HINT="ex: /path/to/joomla/site/administrator/tls/ca/"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_LABEL="CA Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_DESC="The absolute file path to the database client (Joomla server) base64 encoded certificate file (including intermediate certification authorities if applicable)."
```
I have no idea what this means

richard67 replied:
@andrepereiradasilva Do you have an idea for a better text here? I would at least change `database client (Joomla server)` to `database client's`.

andrepereiradasilva replied:
No don't have a better text.

`base64 encoded certificate file` -> base64 is a format for certificates, you can also have binary format

`including intermediate certification authorities if applicable`-> if the database client certificate is signed by an intermediate CA the intermediate CA certificate also needs to be in the file (along with the database client certificate).

richard67 replied:
@brianteeman Is the text better now?

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_DESC="The absolute file path to the database client (Joomla server) base64 encoded certificate file (including intermediate certification authorities if applicable)."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_HINT="ex: /path/to/joomla/site/administrator/tls/certs/my-client-certificate.pem"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_LABEL="Certificate File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="The list of ciphers the database client (Joomla server) allows to use for connection encryption."
```
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="The list of ciphers the database client supports."
```

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_NONE="Default (server controlled)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_ONE_WAY="One-way encryption"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_TWO_WAY="Two-way encryption"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_DESC="The absolute file path to the database client (Joomla server) base64 encoded private key file."
```
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_DESC="The absolute file path."
```

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,28 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_DESC="The absolute file path to the database client (Joomla server) base64 encoded private key file."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_HINT="ex: /path/to/joomla/site/administrator/tls/private/my-client-private-key.pem"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_LABEL="Private Key File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_VERIFY_SERVER_CERT_DESC="Verify the server certificate on client side."
```
remove as it serves no purpose

richard67 replied:
@brianteeman Do you mean this text "COM_CONFIG_FIELD_DATABASE_ENCRYPTION_VERIFY_SERVER_CERT_DESC"? It clarifies that the verification is at client side for the server certificate, the label does not tell that so clearly. Or is it in general that the descriptions are not shown anywhere at the moment? I think I remember in past I had seen them in tool tips, but maybe I remember wrong, and currently I don't see tool tips in Global Config.

brianteeman replied:
what was a tooltip is now displayed underneath the field.

It doesn't really clarify anything and is not needed

brianteeman commented on installation/configuration.php-dist:
```diff
@@ -44,6 +44,13 @@ class JConfig
 	public $password = '';                   // DB password
 	public $db = '';                         // DB database name
 	public $dbprefix = 'jos_';               // Do not change unless you need to!
+	public $dbencryption = 0;
```
all of this needs to be added to installation\src\Model\ConfigurationModel.php

richard67 replied:
@andrepereiradasilva Do we need the configuration.php-dist  in this PR, or have you meant this to be in the PR for adding it to the installation only?

andrepereiradasilva replied:
see https://github.com/richard67/joomla-cms/pull/11/files

richard67 replied:
Thanks @andrepereiradasilva and @brianteeman .

brianteeman commented:
Hints are bad for accessibility for several reasons and are also bad for ux as they disappear as soon as you focus on the input.

If you **absolutely** need the information that you are displaying in the hint then move it to the description. I suspect that you really don't

Ex - is not a valid abbreviation. Use eg 

see https://developer.joomla.org/en-gb-user-interface-text-guidelines/punctuation.html

richard67 committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-Authored-By: Brian Teeman <brian@teeman.net>

richard67 committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-Authored-By: Brian Teeman <brian@teeman.net>

richard67 commented:
@brianteeman Thanks for the feedback and review. I will fix remaining stuff soon.

richard67 committed:
Merge pull request #11 from andrepereiradasilva/patch-34

Need to save in install configuration.php (on install) the new default global config values

richard67 committed:
One CA key file only.

richard67 committed:
Remove hints and move into to descriptions + correct some texts

richard67 commented:
@brianteeman I've removed the hints and added their info to the descriptions. I still have 2 questions, see comments to code above. Could you check again? Thanks in advance and thanks for review up to now.

andrepereiradasilva commented on installation/src/Model/ConfigurationModel.php:
```diff
@@ -133,6 +133,13 @@ public function createConfiguration($options)
 		$registry->set('db', $options->db_name);
 		$registry->set('dbprefix', $options->db_prefix);
+		$registry->set('dbencryption', 0);
+		$registry->set('dbsslverifyservercert', 0);
```
should be "false" here, my mistake

richard67 replied:
Solved. Thanks.

richard67 committed:
Update ConfigurationModel.php

andrepereiradasilva commented on installation/src/Model/ConfigurationModel.php:
```diff
@@ -133,6 +133,13 @@ public function createConfiguration($options)
 		$registry->set('password', $options->db_pass_plain);
 		$registry->set('db', $options->db_name);
 		$registry->set('dbprefix', $options->db_prefix);
+		$registry->set('dbencryption', false);
```
this one is not boolean is 0, 1 or 2 (see the xml)

richard67 replied:
Ok, solved.

richard67 committed:
Grr

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_DOMAIN_DESC="Precede domain with '.' if cookie should be valid for all subdomains."
 COM_CONFIG_FIELD_COOKIE_DOMAIN_LABEL="Cookie Domain"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The absolute file path to the Certificate Authority (CA) base64 encoded certificates file, eg 'ex: /path/to/joomla/site/administrator/tls/ca/my-client-certificate-ca.pem'."
```
For me it is enough to say
"Absolute file path."

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="The absolute file path to the Certificate Authority (CA) base64 encoded certificates file, eg 'ex: /path/to/joomla/site/administrator/tls/ca/my-client-certificate-ca.pem'."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_LABEL="CA File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="The full path to Certificate Authority (CA) base64 encoded certificate files (MySQL only), eg '/path/to/joomla/site/administrator/tls/ca/'. Leave it empty if \"CA File\" already have all the CA certificates needed to validate the chain of trust."
```
If it is mysql only can this field not be displayed conditionally?

be consistent in full path or absolute path

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_LABEL="CA File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="The full path to Certificate Authority (CA) base64 encoded certificate files (MySQL only), eg '/path/to/joomla/site/administrator/tls/ca/'. Leave it empty if \"CA File\" already have all the CA certificates needed to validate the chain of trust."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_LABEL="CA Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_DESC="The absolute file path to the database client's base64 encoded certificate file, eg '/path/to/joomla/site/administrator/tls/certs/my-client-certificate.pem', including any intermediate CA's certificates if used."
```
can it be anything other than base64? If not then you dont need to say that and its enough to simply state
"Absolute file path."

richard67 replied:
Also leave away the example beginning with `eg`?

brianteeman replied:
no need for the example. if you dont know what an absolute file path is then you should read the documentation

andrepereiradasilva replied:
certficates can be in binary format for instance - `base64 encoded` is reggarding the PEM format. https://www.sslshopper.com/ssl-converter.html - not sure if needed here, but there are other formats
the example was to help since people can have issues with, for instance, php open_basedir config, php process needs to be able to reach the ca/certificate/key file - not sure if needed here

brianteeman replied:
that really is something for the documentation not the description.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_LABEL="CA Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_DESC="The absolute file path to the database client's base64 encoded certificate file, eg '/path/to/joomla/site/administrator/tls/certs/my-client-certificate.pem', including any intermediate CA's certificates if used."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_LABEL="Certificate File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="The list of ciphers the database client supports, eg 'EECDH+AESGCM:EDH+AESGCM:EECDH+AES:EDH+AES'."
```
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="Supported ciphers eg 'EECDH+AESGCM:EDH+AESGCM:EECDH+AES:EDH+AES'."
```

richard67 replied:
No comma before `eg`?

brianteeman replied:
no

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CERT_LABEL="Certificate File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_DESC="The list of ciphers the database client supports, eg 'EECDH+AESGCM:EDH+AESGCM:EECDH+AES:EDH+AES'."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CIPHER_LABEL="Cipher Suite"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_DESC="Data in transit encryption (with TLS for database connections)."
```
not needed - you can remove this

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,23 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_NONE="Default (server controlled)"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_ONE_WAY="One-way encryption"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_ENABLE_VALUE_TWO_WAY="Two-way encryption"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_KEY_DESC="The absolute file path to the database client's. private key file, eg '/path/to/joomla/site/administrator/tls/private/my-client-private-key.pem'."
```
Its enough to simply say 
"Absolute file path."

richard67 replied:
Also leave away the example beginning with `eg`?

brianteeman replied:
no need for the example. if you dont know what an absolute file path is then you should read the documentation

richard67 committed:
Simplify language strings and remove some obsolete ones

richard67 committed:
Simplify another one

richard67 committed:
Remove another not needed description

richard67 commented:
@brianteeman Except of the conditional display (showon) for the MySQL-specific field, all should be done as you suggested. I hope I haven't missed or misunderstood anything.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

brianteeman commented:
so will you be doing the conditional display?

richard67 commented:
@brianteeman Am trying. Not sure if `showon="databaseconnection:mysql,mysqli[AND]dbencryption:2"` will work.

richard67 committed:
Display CA path only if MySQL

richard67 commented:
@brianteeman Am testing the showon right now, and if it will not work I'll undo. Do you know if I have to be careful with something regarding showon? Am not very familiar with it.

andrepereiradasilva commented:
> @brianteeman Am testing the showon right now, and if it will not work I'll undo. Do you know if I have to be careful with something regarding showon? Am not very familiar with it.

Should be: `showon="dbtype:mysql,mysqli[AND]dbencryption:2"`

there are actually two fields (cipher and ca) that only work in mysqli adn PDO mysql, not in postgresql
https://github.com/joomla-framework/database/blob/2.0-dev/src/Pdo/PdoDriver.php#L280-L286

richard67 commented:
@andrepereiradasilva Thanks, will change the showon for both. @brianteeman Should I leave the '(MySQL only)' in the text despite of the showon, just in case? Or shall I remove that because showon is enough?

richard67 committed:
Correct showon and text for MySQL only fields

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,21 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="Absolute file path."
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_LABEL="CA File"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="Absolute path. Leave it empty if \"CA File\" already have all the CA certificates needed to validate the chain of trust."
```
```suggestion
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CAPATH_DESC="Absolute file path."
```

richard67 committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-Authored-By: Brian Teeman <brian@teeman.net>

richard67 commented:
@andrepereiradasilva You are right that some information has to be documented somewhere. We have to do that at the help pages or on the docs wiki or both. The `_DESCR` texts are the wrong place for too long explanations, I agree with @brianteeman at this point. Unfortunately I share your fear that this documentation might never happen.

brianteeman commented:
The documentation will happen if you write it - you guys are the best people to write it because you are the ones who understand it the most.

Ideally any new feature should have documentation with it as a requirement before it can be finally merged

richard67 commented:
@brianteeman I agree, but I would not be the ideal writer. @andrepereiradasilva knows the feature better than I do. Maybe we can do that together and I can put it on the wiki. For today I have to finish, gotta work tomorrow morning. And then I have primary focus still on database datetime stuff. Next weekend I could find time, I hope. We stay in contact, Andre.

richard67 commented:
@twister65 and other potential testers: I've updated the patched full install zip package linked in the testing instructions for those who do not have a git repository clone with composer and so on.

PR should be ready for testing now.

brianteeman commented:
@quy please add the documentation required label

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -35,6 +36,21 @@ COM_CONFIG_FIELD_CACHE_TIME_LABEL="Cache Time (minutes)"
 COM_CONFIG_FIELD_COOKIE_DOMAIN_DESC="Precede domain with '.' if cookie should be valid for all subdomains."
 COM_CONFIG_FIELD_COOKIE_DOMAIN_LABEL="Cookie Domain"
 COM_CONFIG_FIELD_COOKIE_PATH_LABEL="Cookie Path"
+COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_DESC="Absolute file path."
```
Having looked again at this afresh this morning I think we can remove the description entirely
so the label would become
COM_CONFIG_FIELD_DATABASE_ENCRYPTION_CA_LABEL="Path to CA File"


If you look at similar fields their label is sufficient
COM_CONFIG_FIELD_CACHE_PATH_LABEL="Path to Cache Folder"
COM_CONFIG_FIELD_LOG_PATH_LABEL="Path to Log Folder"
COM_CONFIG_FIELD_TEMP_PATH_LABEL="Path to Temp Folder"

richard67 replied:
@brianteeman Remove all descriptions added by this PR, or only this one where the comment refers to?

brianteeman replied:
only the ones similar to this one

richard67 replied:
Ok, will implement tonight after work, German time.

richard67 replied:
@brianteeman I've decided to drop all descriptions because it needs documentation anyway to understand the functionality, no description in backend could be long enough to cover all that, and those descriptions still would have been too short (or not precise enough) and so not useful. Are the texts ok now?

brianteeman replied:
Other than comment about break the strings are good now

twister65 commented:
It would be great to add a RFT (Ready For Test) tag for Joomla's PRs :thinking:

richard67 commented:
@twister65 Well, if I had known that it takes several iterations to make it ready for testing, I would have made this PR as draft PR first and then later at the end marked it as ready for review, which means "undraft". But if that happens too early, there is no way back to switch from normal PR back to draft, and so you are again where we are now. This switching back from draft to not draft is a functionality requested by users of GitHub but not impemented yet.

richard67 committed:
Drop useless descriptions and shorten label texts

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -20,6 +20,7 @@ COM_CONFIG_ERROR_CONFIGURATION_PHP_NOTUNWRITABLE="Could not make configuration.p
 COM_CONFIG_ERROR_CONFIGURATION_PHP_NOTWRITABLE="Could not make configuration.php writable."
 COM_CONFIG_ERROR_CUSTOM_CACHE_PATH_NOTWRITABLE_USING_DEFAULT="The folder at %1$s is not writable and cannot be used for the cache, using the default %2$s instead."
 COM_CONFIG_ERROR_CUSTOM_SESSION_FILESYSTEM_PATH_NOTWRITABLE_USING_DEFAULT="The folder at %s is not writable and cannot be used to store session data, the default PHP path will be used instead."
+COM_CONFIG_ERROR_DATABASE_NOT_AVAILABLE="Database connection test failed with the following error: <em>%s: %s</em><br/>Database connection settings changes were not saved."
```
Change to `<br>`

richard67 replied:
`<em>` is ok? Or should I change this to `<strong>`?

richard67 replied:
Ah, I see, both is used here or there in the en-GB.com_config.ini file, so it would not be inconsistent, the `<em>`.

richard67 committed:
Fix language text for HTML5

richard67 commented:
I think texts should be ok now. Thanks @brianteeman for the reviews.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 commented:
@andrepereiradasilva I will try to find time on weekend to write some documentation. If not this weekend, then the next one. I'll ping you for a review at the end before I'll publish, but you shouldn't have much work with it.

richard67 committed:
Merge remote-tracking branch 'upstream/4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 commented:
Solved conflicts.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-tls-encryption-for-db-connections

wilsonge commented:
Thanks!

richard67 commented:
Updated section "Documentation Changes Required" of the description by what I think that is necessary.
